File: 0159-longest-substring-with-at-most-two-distinct-characters/0159-longest-substring-with-at-most-two-distinct-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstringTwoDistinct(string s) {
        int n= s.length();
        if(n == 1 || n == 2) return n;
        
        unordered_map<char, int> umap;
        
        umap[s[0]]++;
        umap[s[1]]++;
        
        int ans = 0;
        
        for(int i=2, j=0; i<n; i++){
            umap[s[i]]++;
            for(; j<i && umap.size() > 2; j++){
                if(umap[s[j]] == 1){
                    umap.erase(s[j]);
                } else{
                    umap[s[j]]--;
                }
            }  
            ans = max(ans, i-j+1);
        }
            
        
        return ans;
    }
};

File: 0159-longest-substring-with-at-most-two-distinct-characters/0159-longest-substring-with-at-most-two-distinct-characters.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        int n = s.length();
        if(n == 1 || n == 2) return n;
        
        Map<Character,Integer> umap = new HashMap<>();
                
        int ans = 0;
        
        char []ch = s.toCharArray();
        
        for(int i=0, j=0; i<n; i++){
            umap.put(ch[i], umap.getOrDefault(ch[i], 0) + 1);
            for(; j<i && umap.size() > 2; j++){
                if(umap.get(ch[j]) == 1){
                    umap.remove(ch[j]);
                } else{
                    umap.put(ch[j], umap.get(ch[j]) - 1);
                }
            }  
            ans = Math.max(ans, i-j+1);
        }
            
        
        return ans;
    }
}